File: Manufacturing/src/Python/package/microsoft_azure_sphere_deviceapi/certificate.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from os import path
from typing import Literal

from microsoft_azure_sphere_deviceapi import utils
from microsoft_azure_sphere_deviceapi.exceptions import ValidationError

__all__ = ['add_certificate', 'get_all_certificates',
           'get_certificate', 'get_certificate_space', 'remove_certificate']


def add_certificate(
    certificate_id: str,
    certificate_path: str,
    cert_type: Literal["client", "rootca"],
    private_key_path="",
    password="",
) -> dict:
    """Makes a "POST" request to add a certificate to the attached devices certificate store.

    :param certificate_id: The ID of the certificate.
    :type certificate_id: string
    :param certificate_path: The absolute path to a public key certificate .pem file.
    :type certificate_path: string
    :param cert_type: The type of certificate to add.
    :type cert_type: string
    :accepted-values cert_type: client, rootca
    :param private_key_path: The absolute path to a client private key .pem file. Required when  
                             adding a certificate of type 'client'.
    :type private_key_path: string
    :param password: The password to decrypt the private key. Required when adding a
                     client private key that is encrypted.
    :type password: string

    :return: An empty response on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    if not certificate_id:
        raise ValidationError("ERROR: Cannot add certificate, certificate ID is null or empty.")

    if not certificate_path:
        raise ValidationError(
            "ERROR: Cannot add certificate, certificate path is null or empty.")

    valid_cert_types = ["client", "rootca"]

    if not cert_type or cert_type not in valid_cert_types:
        raise ValidationError("ERROR: Cannot add certificate, cert type is invalid.")

    if not path.exists(certificate_path) or not path.isfile(certificate_path):
        raise ValidationError("ERROR: Cannot add certificate, certificate path is invalid.")

    cert = open(certificate_path, "r")

    response = None
    if cert_type == "client":
        if not private_key_path:
            raise ValidationError("ERROR: Cannot add certificate, private key path is invalid.")

        if not password:
            raise ValidationError("ERROR: Cannot add certificate, invalid password.")

        private_key_file = open(private_key_path, "r")
        response = utils.post_request(
            f"certstore/certs/{certificate_id}",
            {
                "certType": cert_type,
                "publicCert": cert.read(),
                "privateKey": private_key_file.read(),
                "password": password,
            },
        )
        private_key_file.close()
    else:
        response = utils.post_request(
            f"certstore/certs/{certificate_id}",
            {"certType": cert_type, "publicCert": cert.read()},
        )
    cert.close()
    return response


def get_all_certificates() -> dict:
    """List certificates in the attached device's certificate store.

    :return: The identifiers of the attached certificates on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("certstore/certs")


def get_certificate(certificate_id: str) -> dict:
    """Makes a REST "GET" request to get information about a specific certificate.

    :param certificate_id: The ID of the certificate.
    :type certificate_id: string

    :return: Details of the certificate on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    if not certificate_id:
        raise ValidationError("ERROR: Cannot get certificate, certificate ID is null or empty.")

    return utils.get_request(f"certstore/certs/{certificate_id}")


def get_certificate_space() -> dict:
    """Makes a REST "GET" request to show the available free space in the attached device's
    certificate store.

    :return:The amount of free space for certificates on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("certstore/space")


def remove_certificate(certificate_id: str) -> dict:
    """Makes a "DELETE" request to delete a certificate in the attached device's certificate store.

    :param certificate_id: The ID of the certificate.
    :type certificate_id: string

    :return: An empty response on success. An exception will be thrown on error.

    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    if not certificate_id:
        raise ValidationError("ERROR: Cannot get certificate, certificate ID is null or empty.")

    return utils.delete_request(f"certstore/certs/{certificate_id}")

File: Manufacturing/src/CSharp/Nuget/Package/Certificate.cs
/* Copyright (c) Microsoft Corporation. All rights reserved.
   Licensed under the MIT License. */
namespace Microsoft.Azure.Sphere.DeviceAPI
{
    using System.IO;
    using System;

    /// <summary>
    /// Device REST APIs to manage certificates on the attached device. 
    /// </summary>
    public static class Certificate
    {
        /// <summary>Makes a "POST" request to add a certificate to an attached device certificate store.</summary>
        /// <param name="certificateID">The ID of the certificate.</param>
        /// <param name="certificatePath">The path of your certificate file.</param>
        /// <param name="certType">The type of certificate to add.</param>
        /// <param name="privateKeyLocation">The location of your private key file.</param>
        /// <param name="password">The password to decrypt the private key.</param>
        /// <remarks>Accepted certType values: ["client", "rootca"]</remarks>
        /// <returns>An empty response a string on success. An exception will be thrown on error.</returns>
        public static string AddCertificate(string certificateID, string certificatePath, string certType, string privateKeyLocation = "", string password = "")
        {
            if (string.IsNullOrEmpty(certificatePath))
            {
                throw new ValidationError("Cannot add certificate, certificate location is null or empty.");
            }

            if (!File.Exists(certificatePath))
            {
                throw new ValidationError("Cannot add certificate, file doesn't exist or cannot open.");
            }

            if (string.IsNullOrEmpty(certificateID))
            {
                throw new ValidationError("Cannot add certificate, certificate ID is null or empty.");
            }

            string[] validCertTypes = new string[] { "client", "rootca" };

            if (string.IsNullOrEmpty(certType) || !Array.Exists(validCertTypes, elem => elem.Equals(certType)))
            {
                throw new ValidationError("Cannot add certificate, certificate type is invalid.");
            }

            string publicCert = File.ReadAllText(certificatePath);
            string response = String.Empty;
            string urlWithComponent = $"certstore/certs/{certificateID}";

            if (certType.Equals("client"))
            {

                if (string.IsNullOrEmpty(privateKeyLocation))
                {
                    throw new ValidationError("Cannot add certificate, private key location is null or empty.");
                }

                if (string.IsNullOrEmpty(password))
                {
                    throw new ValidationError("Cannot add certificate, password is null or empty.");
                }

                string privateKey = File.ReadAllText(privateKeyLocation);
                response = RestUtils.PostRequest(urlWithComponent, new { certType, publicCert, privateKey, password });
            }
            else
            {
                response = RestUtils.PostRequest(urlWithComponent, new { certType, publicCert });
            }
            return response;
        }

        /// <summary>Makes a REST "GET" request to retrieve a list of certificates from an attached device certificate store.</summary>
        /// <returns>The identifiers of the attached certificates as a string on success. An exception will be thrown on error.</returns>
        public static string GetAllCertificates()
        {
            return RestUtils.GetRequest("certstore/certs");
        }

        /// <summary>Makes a REST "GET" request to retrieve information about a specific certificate from an attached device.</summary>
        /// <param name="certificateID"> The ID of the certificate.</param>
        /// <returns>Details of the certificate as a string on success. An exception will be thrown on error.</returns>
        public static string GetCertificate(string certificateID)
        {
            if (string.IsNullOrEmpty(certificateID))
            {
                throw new ValidationError("Cannot Remove Certificate, certificate ID is null or empty.");
            }

            return RestUtils.GetRequest($"certstore/certs/{certificateID}");
        }

        /// <summary>Makes a REST "GET" request to retrieve the available free space for an attached device certificate store.</summary>
        /// <returns>The amount of free space for certificates as a string on success. An exception will be thrown on error.</returns>
        public static string GetCertificateSpace()
        {
            return RestUtils.GetRequest("certstore/space");
        }

        /// <summary>Makes a "DELETE" request to delete a certificate in an attached device certificate store.</summary>
        /// <param name="certificateID">The ID of the certificate.</param>
        /// <returns>An empty response as a string on success. An exception will be thrown on error.</returns>
        public static string RemoveCertificate(string certificateID)
        {
            if (string.IsNullOrEmpty(certificateID))
            {
                throw new ValidationError("Cannot remove certificate, certificate ID is null or empty.");
            }
            return RestUtils.DeleteRequest($"certstore/certs/{certificateID}");
        }
    }

    /// <summary>Helper class, providing options for AddCertificate</summary>
    public static class AddCertificateCertTypeOptions
    {
        /// <summary>client option</summary>
        public static string client = "client";
        /// <summary>rootca option</summary>
        public static string rootca = "rootca";
    }
}
